onmt/train/Trainer.lua

```diff
@@ -254,10 +271,24 @@ function Trainer:train(model, optim, trainData, validData, dataset, info)
       end
     end
 
+    -- if needed, get avgPpl from epochState before log() since log() will reset internal stats
+    local avgPpl
+    if self.options.sample and self.options.sample_w_ppl then
+      avgPpl = epochState:getAvgPpl()
```

Comment:
Reverted with 8c3fcf3. I didn't know epochState keeps only the stat from last batch, The module name implies otherwise.